src/librustc_resolve/resolve_imports.rs

```diff
@@ -867,12 +868,58 @@ impl<'a, 'b:'a> ImportResolver<'a, 'b> {
             }
 
             match binding.kind {
-                NameBindingKind::Import { binding: orig_binding, .. } => {
+                NameBindingKind::Import { binding: orig_binding, directive, .. } => {
                     if ns == TypeNS && orig_binding.is_variant() &&
-                       !orig_binding.vis.is_at_least(binding.vis, &*self) {
-                        let msg = format!("variant `{}` is private, and cannot be reexported, \
-                                           consider declaring its enum as `pub`", ident);
-                        self.session.span_err(binding.span, &msg);
+                        !orig_binding.vis.is_at_least(binding.vis, &*self) {
+                            let msg = match directive.subclass {
+                                ImportDirectiveSubclass::SingleImport { .. } => {
+                                    "variant is private and cannot be reexported".to_owned()
+                                },
+                                ImportDirectiveSubclass::GlobImport { .. } => {
+                                    let msg = "enum is private and its variants \
+                                               cannot be reexported".to_owned();
+                                    let error_id = (DiagnosticMessageId::ErrorId(0), // no code?!
```

Comment:
You can create a new code in https://github.com/rust-lang/rust/blob/master/src/librustc_resolve/diagnostics.rs (probably a good idea to do it as part of this PR)

Reply:
I think there's a strong case to be made that this should just be [E0364](https://doc.rust-lang.org/error-index.html#E0364), rather than a new code.

Reply:
I'd rather defer this to some future PR. (I [started coding](https://github.com/zackmdavis/rust/commit/ef65c11fc), but the pieces don't quite fit together for the elegant thing that I wanted to do.)